packages/tangle/analysis.go

```diff
@@ -0,0 +1,231 @@
+
+func (a ByArrival) Len() int { return len(a) }
+func (a ByArrival) Less(i, j int) bool {
+	return a[i].MsgArrivalTime.Before(a[j].MsgArrivalTime)
+}
+func (a ByArrival) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+// BySolid defines a slice of MsgInfo sortable by solid time.
+type BySolid []MsgInfo
+
+func (a BySolid) Len() int { return len(a) }
+func (a BySolid) Less(i, j int) bool {
+	return a[i].MsgSolidTime.Before(a[j].MsgSolidTime)
+}
+func (a BySolid) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+
+// FirstApprovalAnalysis performs the first approval analysis and write the result into a csv.
+// This function is very heavy to compute especially when starting from the Genesis.
+// A better alternative for the future would be to keep this analysis updated as the Tangle grows.
+func (t *Tangle) FirstApprovalAnalysis(nodeID string, filePath string) error {
+	// If the file doesn't exist, create it, or truncate the file
+	f, err := os.Create(filePath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	w := csv.NewWriter(f)
+
+	// write TableDescription
+	if err := w.Write(TableDescription); err != nil {
+		return err
+	}
+
+	return t.FutureCone(EmptyMessageID, func(msgID MessageID) error {
+		approverInfo, err := t.firstApprovers(msgID)
+		// firstApprovers returns an error when the msgID is a tip, thus
+		// we want to stop the computation but continue with the future cone iteration.
+		if err != nil {
+			return nil
+		}
+
+		msgApproval := MsgApproval{
+			NodeID:                  nodeID,
+			Msg:                     t.info(msgID),
+			FirstApproverByIssuance: approverInfo[byIssuance],
+			FirstApproverByArrival:  approverInfo[byArrival],
+			FirstApproverBySolid:    approverInfo[bySolid],
+		}
+
+		// write msgApproval to file
+		if err := w.Write(msgApproval.toCSV()); err != nil {
+			return err
+		}
+		w.Flush()
+		if err := w.Error(); err != nil {
+			return err
+		}
+		return nil
+	})
+}
+
+// FutureCone iterates over the future cone of the given messageID and computes the given function.
+func (t *Tangle) FutureCone(messageID MessageID, compute func(ID MessageID) error) error {
+	futureConeStack := list.New()
+	futureConeStack.PushBack(messageID)
+
+	processedMessages := make(map[MessageID]types.Empty)
+	processedMessages[messageID] = types.Void
+
+	for futureConeStack.Len() >= 1 {
+		currentStackEntry := futureConeStack.Front()
+		currentMessageID := currentStackEntry.Value.(MessageID)
+		futureConeStack.Remove(currentStackEntry)
+
+		if err := compute(currentMessageID); err != nil {
+			return err
+		}
+
+		t.Approvers(currentMessageID).Consume(func(approver *Approver) {
+			approverID := approver.ApproverMessageID()
+			if _, messageProcessed := processedMessages[approverID]; !messageProcessed {
+				futureConeStack.PushBack(approverID)
+				processedMessages[approverID] = types.Void
+			}
+		})
+	}
+
+	return nil
+}
+
+func (t *Tangle) firstApprovers(msgID MessageID) ([]MsgInfo, error) {
+	approversInfo := []MsgInfo{}
```

Comment:
my linter complains about this
```suggestion
	approversInfo := make([]MsgInfo, 0)
```